### ISSUE TITLE
Interrupted decomission test hangs with tablets

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#17027]
In  #17026 we saw that the dtest `materialized_views_test.py`::`TestMaterializedViews`::`test_mv_resurrected_rows_after_decommission_interrupt`[`without_rbno`] (and also `with_rbno`) fail when consistent topology changes is enabled, but the failure is minor - it's because of a spurious ERROR log message.

When additionally, **tablets** are enabled, the failure becomes different and worse - the interrupted decomissioned node seems to hang in a weird way. We see in the log of the 4th node:

```
INFO  2024-01-28 23:28:25,853 [shard 0:strm] storage_service - DECOMMISSIONING: done
```
But then Scylla hangs for 90 seconds until the message:
```
WARN  2024-01-28 23:29:59,757 [shard 0:main] service_level_controller - update_from_distributed_data: failed to update configuration for more than  90 seconds : exceptions::read_timeout_exception (Operation timed out for system_distributed.service_levels - received only 0 responses from 1 CL=ONE.)
```
followed by repeats of this message with increasing timeouts until:
```
WARN  2024-01-28 23:37:19,760 [shard 0:main] service_level_controller - update_from_distributed_data: failed to update configuration for more than  540 seconds : exceptions::read_timeout_exception (Operation timed out for system_distributed.service_levels - received only 0 responses from 1 CL=ONE.)
```
After such a long time, dtest gives up and kills Scylla, and the log shows a normal shutdown
```
INFO  2024-01-28 23:38:39,425 [shard 0:main] init - Signal received; shutting down
...
```

**Comment by bhalevy** [comment]
Cc @tgrabiec

**Comment by tgrabiec** [comment]
@nyh Do you have full logs from this run?

**Comment by nyh** [comment]
In another run, the same two tests `materialized_views_test.py::TestMaterializedViews::test_mv_resurrected_rows_afte
r_decommission_interrupt[with_rbno]` and `[without_rbno]` both failed differently - it didn't hang and instead complained about not getting the expected error from nodetool:

```
        with ThreadPoolExecutor(max_workers=1) as decommission_thread:
>           with pytest.raises(expected_exception=NodetoolError):
E           Failed: DID NOT RAISE <class 'ccmlib.node.ToolError'>

materialized_views_test.py:823: Failed
```

**Comment by nyh** [comment]
> @nyh Do you have full logs from this run?

[logs.tar.gz](https://github.com/scylladb/scylladb/files/14085063/logs.tar.gz)

**Comment by tgrabiec** [comment]
It's a test problem. With tablets, InterruptDecommission never matches expected strings because it expects `Unbootstrap with %s for keyspace=ks succeeded`, which is only printed for vnode-based streaming, it is not printed for tablet-based keyspaces which are drained in the tablet migration track. 

```
grep "Unbootstrap with" node4.log  | grep succee
INFO  2024-01-29 03:12:49,264 [shard 0: gms] range_streamer - Unbootstrap with 127.0.70.3 for keyspace=system_distributed succeeded, took 0.016285617 seconds
INFO  2024-01-29 03:12:49,265 [shard 0: gms] range_streamer - Unbootstrap with 127.0.70.1 for keyspace=system_distributed succeeded, took 0.01687664 seconds
INFO  2024-01-29 03:12:49,265 [shard 0: gms] range_streamer - Unbootstrap with 127.0.70.2 for keyspace=system_distributed succeeded, took 0.01687664 seconds
INFO  2024-01-29 03:12:49,275 [shard 0: gms] range_streamer - Unbootstrap with 127.0.70.1 for keyspace=system_traces succeeded, took 0.010005466 seconds
INFO  2024-01-29 03:12:49,276 [shard 0: gms] range_streamer - Unbootstrap with 127.0.70.3 for keyspace=system_traces succeeded, took 0.010685655 seconds
INFO  2024-01-29 03:12:49,276 [shard 0: gms] range_streamer - Unbootstrap with 127.0.70.2 for keyspace=system_traces succeeded, took 0.01095563 seconds
INFO  2024-01-29 03:12:49,282 [shard 0: gms] range_streamer - Unbootstrap with 127.0.70.2 for keyspace=system_auth succeeded, took 0.006613894 seconds
INFO  2024-01-29 03:12:49,282 [shard 0: gms] range_streamer - Unbootstrap with 127.0.70.3 for keyspace=system_auth succeeded, took 0.006696839 seconds
INFO  2024-01-29 03:12:49,283 [shard 0: gms] range_streamer - Unbootstrap with 127.0.70.1 for keyspace=system_auth succeeded, took 0.007046165 seconds
```

So the test hangs in interrupt_decommission.join()

**Comment by tgrabiec** [comment]
> In another run, the same two tests `materialized_views_test.py::TestMaterializedViews::test_mv_resurrected_rows_afte r_decommission_interrupt[with_rbno]` and `[without_rbno]` both failed differently - it didn't hang and instead complained about not getting the expected error from nodetool:
> 
> ```
>         with ThreadPoolExecutor(max_workers=1) as decommission_thread:
> >           with pytest.raises(expected_exception=NodetoolError):
> E           Failed: DID NOT RAISE <class 'ccmlib.node.ToolError'>
> 
> materialized_views_test.py:823: Failed
> ```

Was it with tablets?

**Comment by tgrabiec** [comment]
> If you're familiar with that "Unboostrap" stuff, do you by any chance know which if there's a different message which replaced this message?

The test seems to wait until all keyspaces are streamed, so maybe it's better to wait for one message printed after everything.

There is now error injection point after streaming in decommission: `storage_service_streaming_sleep3`, maybe we should plug into it (!release mode only).

**Comment by nyh** [comment]
The problem here is probably not specific to an **interrupted** decomission, and every decommission hangs (at least in the context of MV tests) - I opened https://github.com/scylladb/scylladb/issues/17293.
Perhaps this issue should later be closed as a messier duplicate of that other one.

**Comment by tgrabiec** [comment]
This was already diagnosed to be a test problem, no? https://github.com/scylladb/scylladb/issues/17027#issuecomment-1915761837

**Comment by nyh** [comment]
> This was already diagnosed to be a test problem, no? [#17027 (comment)](https://github.com/scylladb/scylladb/issues/17027#issuecomment-1915761837)

I think it might be several independent problems which is why I didn't yet rush to close this issue. In https://github.com/scylladb/scylladb/issues/17293 I show that a straight-up decomission  - without interrupting it, never finishes - the decommission seems to finish but the node never shuts down and instead hangs in some "service levels" related loop. So any explanation that is specific to how interrupt_decommission works cannot be the whole story.

**Comment by xemul** [comment]
> I think it might be several independent problems which is why I didn't yet rush to close this issue. In #17293 I show that a straight-up decomission - without interrupting it, never finishes

Correction -- decommission finishes. It's _test_ that waits for views to be built and this never happens

**Comment by xemul** [comment]
> This was already diagnosed to be a test problem, no? [#17027 (comment)](https://github.com/scylladb/scylladb/issues/17027#issuecomment-1915761837)

Yes, that's the problem. The test waits for the `Unbootstrap with <address> for keyspace=ks succeeded`, but keyspace ks is not "unbootstrapped" in this sense, there's no such line in logs, keyspace `ks` is migrated away by other means:

```
stream_session - [Stream #6aff8c70-d659-11ee-a610-49f753bea305] Streaming plan for Tablet migration-ks-index-0 succeeded, peers={<address>}
```

**Comment by xemul** [comment]
:point_up: the dtest fix above makes the non-rbno run pass

**Comment by xemul** [comment]
Both changes have hit their masters
